[SDK-4125] [SDK-4578] Flutter deep link set up guide (#7267)
braze-inc/braze-docs: _docs/_developer_guide/platform_integration_guides/flutter/linking.md

You can see [an example app][5] with all of these steps implemented in our public repository. For general information on what deep links are, refer to our [FAQ article][1].

## Step 1: Native deep link handling

Integrating deep links into a Flutter app requires setting up native layer deep links as a pre-requisite. You can follow our guides for setting up deep links on [Android][2] and [iOS][3].

## Step 2: Native project configuration.

If you intend to use Flutter's default deep link handling, you will need to modify your iOS project's `Info.plist` file as well as your Android project's `AndroidManifest.xml` file. If you plan on using third-party plugins to handle deep links in Flutter instead of the default handler, skip this step as it can disrupt those third-party plugins. In the example code provided later in this guide, we will not be using third-party plugins.

{% tabs %}
{% tab iOS %}

> Using Xcode to edit your `Info.plist` file:

Reviewer: ```suggestion
Using Xcode, edit your `Info.plist` file:
```

1. Add a new key-value pair.
2. Set the key to `FlutterDeepLinkingEnabled`.
3. Set the type to `Boolean`.
4. Set the value to `YES`.

![An example project's `Info.plist` file with the added key-value pair.]({% image_buster /assets/img/flutter/flutter-ios-deep-link-info-plist.png %} "Xcode Project Info.plist File")
{% endtab %}

{% tab Android %}
In your `AndroidManifest.xml` file:
1. Locate your `.MainActivity` `activity` tag.
2. Inside of that `activity` tag, add the following `meta-data` tag:

```xml
  <meta-data android:name="flutter_deeplinking_enabled" android:value="true" />
```
{% endtab %}
{% endtabs %}

## Step 3: Dart layer deep link handling and native layer deep link forwarding

The preceding step is sufficient for enabling your Flutter app to open when a user clicks a deep link. For handling deep links in more intricate ways, such as navigating to a particular part of the app or calling a function, additional Dart code as well as additional native code may be necessary. This is where first-party packages such as [`go_router`][6] as well as third-party plugins can be helpful.

> To keep this guide dependency-agnostic, we'll show an example that does not use additional dependencies. In this example, we'll handle deep links by presenting an alert modal after the app launches.

Reviewer: Question to the docs team: Are there any concerns surrounding written voice when using `I` and `we`?

---

# Deep linking

> Deep linking is a way of providing a link that launches an app, shows specific content, or takes some specific action. Native code can forward deep links to your app's Flutter layer for handling. If you're looking to implement deep links in your iOS and/or Android Flutter app for the first time, follow these steps.

For general information on what deep links are, refer to our [FAQ article][1]. 

## Step 1: Native deep link handling

Integrating deep links into a Flutter app requires setting up native layer deep links as a pre-requisite. You can follow our guides for setting up deep links on [Android][2] and [iOS][3].

## Step 2: Native project configuration.

If you intend to use Flutter's default deep link handling, you will need to modify your iOS project's `Info.plist` file as well as your Android project's `AndroidManifest.xml` file. If you plan on using third-party plugins to handle deep links in Flutter instead of the default handler, skip this step as it can disrupt those third-party plugins. In the example code provided later in this guide, we will not be using third-party plugins.

{% tabs %}
{% tab iOS %}

> Using Xcode to edit your `Info.plist` file, add a new key-value pair. Set the key to `FlutterDeepLinkingEnabled`, the type to `Boolean`, and the value to `YES`.

Reviewer 2: Optional suggestion: Maybe a list similar to this would improve readability for the steps here? - 
```suggestion
Using Xcode to edit your `Info.plist` file:
1. Add a new key-value pair.
2. Set the key to `FlutterDeepLinkingEnabled`.
3. Set the type to `Boolean`.
4. Set the value to `YES`.
```

![An example project's `Info.plist` file with the added key-value pair.]({% image_buster /assets/img/flutter/flutter-ios-deep-link-info-plist.png %} "Xcode Project Info.plist File")
{% endtab %}

{% tab Android %}
Add the following `meta-data` and `intent-filter` tags to your `.MainActivity` `activity` tag.

```xml
<activity android:name=".MainActivity"...>
  <meta-data android:name="flutter_deeplinking_enabled" android:value="true" />
  <intent-filter android:autoVerify="true">
      <action android:name="android.intent.action.VIEW" />
      <category android:name="android.intent.category.DEFAULT" />
      <category android:name="android.intent.category.BROWSABLE" />
      <data android:scheme="http" android:host="example.com" />
      <data android:scheme="https" />
  </intent-filter>
</activity>
```
{% endtab %}
{% endtabs %}

> ## Step 3: Flutter layer deep link handling and native layer deep link forwarding

Reviewer 2: Nit:
```suggestion
## Step 3: Dart layer deep link handling and native layer deep link forwarding
```

> The preceeding step is sufficient for enabling your Flutter app to open when a user clicks a deep link. For handling deep links in more intricate ways, such as navigating to a particular part of the app or calling a function, additional Dart code as well as additional native code may be necessary. This is where first-party packages such as `go_router` as well as third-party plugins can be helpful.

Reviewer 2: Nit:
```suggestion
The preceding step is sufficient for enabling your Flutter app to open when a user clicks a deep link. For handling deep links in more intricate ways, such as navigating to a particular part of the app or calling a function, additional Dart code as well as additional native code may be necessary. This is where first-party packages such as `go_router` as well as third-party plugins can be helpful.
```

Reviewer 2: We might also want to add a hyperlink here to [go_router](https://pub.dev/packages/go_router) so customers can read more about what it is

    handleDeepLink(intent)
  }

  private fun handleDeepLink(intent: Intent) {
    val binaryMessenger = flutterEngine?.dartExecutor?.binaryMessenger
    if (intent?.action == Intent.ACTION_VIEW && binaryMessenger != null) {
      MethodChannel(binaryMessenger, "deepLinkChannel")
        .invokeMethod("receivedDeepLink", intent?.data.toString())
    }
  }

}
```

{% endtab %}
{% endtabs %}

Now, in your Dart code, establish the corresponding channel and set the callback to do whatever URL handling you want. In this example, we'll display an alert containing the text of the URL.

```dart
MethodChannel('deepLinkChannel').setMethodCallHandler((call) async {
  deepLinkAlert(call.arguments, context);
});

void deepLinkAlert(String link, BuildContext context) {
  showDialog(
    context: context,
    builder: (BuildContext context) {
      return AlertDialog(
        title: Text("Deep Link Alert"),
        content: Text("Opened with deep link: $link"),
        actions: <Widget>[
          TextButton(
            child: Text("Close"),
            onPressed: () {
              Navigator.of(context).pop();
            },
          ),
        ],
      );
    },
  );
}
```

> 

Reviewer 2: Optional suggestion: We may also want to add a note at the bottom to check out our Sample App for the full example

Author: Ah yeah that's a really good idea actually. We should definitely be leveraging that here.
